src/module_3/module_3.md

```diff
@@ -0,0 +1,891 @@
+for c in cat_cols:
+    train_freq = X_train[c].value_counts()
+    X_train[c] = X_train[c].map(train_freq)
+    
+    test_freq = X_test[c].value_counts()
+    X_test[c] = X_test[c].map(test_freq)
+```
+
+
+```python
+sc = StandardScaler()
+
+X_train_scaled = sc.fit_transform(X_train)
+X_test_scaled = sc.transform(X_test)
+```
+
+
+```python
+from sklearn.model_selection import GridSearchCV
+from sklearn.metrics import classification_report
+from sklearn.model_selection import cross_validate
+from imblearn.over_sampling import SMOTE
+
+
+sgd_log = SGDClassifier(loss='log_loss',class_weight='balanced')
+sgd_svm = SGDClassifier(loss='hinge', class_weight='balanced')
+ridge = RidgeClassifier(class_weight='balanced')
+log_reg = LogisticRegression(class_weight='balanced')
+
+oversample = SMOTE()
+X_smote, y_smote = oversample.fit_resample(X_train_scaled, y_train)
+
+
+
+models = [('SGD LOG', sgd_log),
+          ('SGD SVM', sgd_svm),
+          ('Ridge', ridge),
+          ('Logistic Regression', log_reg)]
+
+accuracy = []
+precision = []
+recall = []
+names = []
+
+
+for name, model in models:
+    scores = cross_validate(model, X_smote, y_smote, cv=10, scoring=['accuracy', 'precision', 'recall'])
+    accuracy.append(scores['test_accuracy'].mean())
+    precision.append(scores['test_precision'].mean())
+    recall.append(scores['test_recall'].mean())
+    names.append(name)
+
+
+results = pd.DataFrame({'Model': names, 'Accuracy': accuracy, 'Precision': precision, 'Recall': recall})
+
+results
+```
+
+
+
+
+<div>
+<style scoped>
+    .dataframe tbody tr th:only-of-type {
+        vertical-align: middle;
+    }
+
+    .dataframe tbody tr th {
+        vertical-align: top;
+    }
+
+    .dataframe thead th {
+        text-align: right;
+    }
+</style>
+<table border="1" class="dataframe">
+  <thead>
+    <tr style="text-align: right;">
+      <th></th>
+      <th>Model</th>
+      <th>Accuracy</th>
+      <th>Precision</th>
+      <th>Recall</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th>0</th>
+      <td>SGD LOG</td>
+      <td>0.739644</td>
+      <td>0.800940</td>
+      <td>0.637845</td>
+    </tr>
+    <tr>
+      <th>1</th>
+      <td>SGD SVM</td>
+      <td>0.736903</td>
+      <td>0.807866</td>
+      <td>0.621708</td>
+    </tr>
+    <tr>
+      <th>2</th>
+      <td>Ridge</td>
+      <td>0.712254</td>
+      <td>0.804072</td>
+      <td>0.561273</td>
+    </tr>
+    <tr>
+      <th>3</th>
+      <td>Logistic Regression</td>
+      <td>0.739791</td>
+      <td>0.800841</td>
+      <td>0.638327</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+</p>As whe have class imbalance, we use `class_weight='balance'` and oversample to increase the number of samples in the underrepresented class.</p>
+<p>Based on the previous cells, we stick with Logistic Regression that has a higher score across accuracy, precision and recall. Although, due to the class imbalance the metric we are most interested in is recall.</p>
```

Comment:
At what threshold is this precision / recall? Why are you using this threshold and why it's important for the business? Could you use a PR curve instead to inform to the business of what would happen with different thresholds?